lib/Mojolicious/Guides/Tutorial.pod

```diff
@@ -0,0 +1,938 @@
+
+  # Multipart upload handler
+  post '/upload' => sub ($c) {
+
+    # Check file size
+    return $c->render(text => 'File is too big.', status => 200) if $c->req->is_limit_exceeded;
+
+    # Process uploaded file
+    return $c->redirect_to('form') unless my $example = $c->param('example');
+    my $size = $example->size;
+    my $name = $example->filename;
+    $c->render(text => "Thanks for uploading $size byte file $name.");
+  };
+
+  app->start;
+  __DATA__
+
+  @@ form.html.ep
+  <!DOCTYPE html>
+  <html>
+    <head><title>Upload</title></head>
+    <body>
+      %= form_for upload => (enctype => 'multipart/form-data') => begin
+        %= file_field 'example'
+        %= submit_button 'Upload'
+      % end
+    </body>
+  </html>
+
+To protect you from excessively large files there is also a limit of 16MiB by default, which you can tweak with the
+attribute L<Mojolicious/"max_request_size">.
+
+  # Increase limit to 1GiB
+  app->max_request_size(1073741824);
+
+=head2 User agent
+
+With L<Mojo::UserAgent>, which is available through the helper L<Mojolicious::Plugin::DefaultHelpers/"ua">, there's a
+full featured HTTP and WebSocket user agent built right in. Especially in combination with L<Mojo::JSON> and
+L<Mojo::DOM> this can be a very powerful tool.
+
+  use Mojolicious::Lite -signatures;
+
+  # Blocking
+  get '/headers' => sub ($c) {
+    my $url = $c->param('url') || 'https://mojolicious.org';
+    my $dom = $c->ua->get($url)->result->dom;
+    $c->render(json => $dom->find('h1, h2, h3')->map('text')->to_array);
+  };
+
+  # Non-blocking
+  get '/title' => sub ($c) {
+    $c->ua->get('mojolicious.org' => sub ($ua, $tx) {
+      $c->render(data => $tx->result->dom->at('title')->text);
+    });
+  };
+
+  # Concurrent non-blocking
+  get '/titles' => sub ($c) {
+    my $mojo = $c->ua->get_p('https://mojolicious.org');
+    my $cpan = $c->ua->get_p('https://metacpan.org');
+    Mojo::Promise->all($mojo, $cpan)->then(sub ($mojo, $cpan) {
+      $c->render(json => {
+        mojo => $mojo->[0]->result->dom->at('title')->text,
+        cpan => $cpan->[0]->result->dom->at('title')->text
+      });
+    })->wait;
+  };
+
+  app->start;
+
+For more information about the user agent see also L<Mojolicious::Guides::Cookbook/"USER AGENT">.
+
+=head2 WebSockets
+
+WebSocket applications have never been this simple before. Just receive messages by subscribing to events such as
+L<Mojo::Transaction::WebSocket/"json"> with L<Mojolicious::Controller/"on"> and return them with
+L<Mojolicious::Controller/"send">.
+
+  use Mojolicious::Lite -signatures;
+
+  websocket '/echo' => sub ($c) {
+    $c->on(json => sub ($c, $hash) {
+      $hash->{msg} = "echo: $hash->{msg}";
+      $c->send({json => $hash});
+    });
+  };
+
+  get '/' => 'index';
+
+  app->start;
+  __DATA__
+
+  @@ index.html.ep
+  <!DOCTYPE html>
+  <html>
+    <head>
+      <title>Echo</title>
+      <script>
+        const ws = new WebSocket('<%= url_for('echo')->to_abs %>');
```

Comment:
Would be really good to include (like jquery) or at least recommend using ReconnectingWebsocket -- unless I'm missing something fundamental.  IME following examples, the websocket will connect and disconnect on its own within a minute or two (despite setting inactivity_timeout to an hour).  After tweaking all timeouts, it seems like the only solution is ReconnectingWebsocket.  Moreover, even with an adequate timeout, apps usually don't want their websocket disconnecting and not reconnecting.  Or if ReconnectingWebsocket is too much activity to the server, what is the recommended way to assure a client side connection establishes a websocket when needed?